cli/tools/registry/pm.rs

```diff
@@ -38,313 +28,210 @@ use crate::file_fetcher::FileFetcher;
-      Ok(Some(Self {
-        dependencies: pkg_json.dependencies.clone().unwrap_or_default(),
-        dev_dependencies: pkg_json.dev_dependencies.clone().unwrap_or_default(),
-        config: pkg_json.clone(),
-        fmt_options: None,
-      }))
-    } else {
-      Ok(None)
-    }
+  fn contents(&self) -> String {
+    self.cst.to_string()
   }
 
   fn add(&mut self, selected: SelectedPackage, dev: bool) {
-    let (name, version) = package_json_dependency_entry(selected);
-    if dev {
-      self.dependencies.swap_remove(&name);
-      self.dev_dependencies.insert(name, version);
-    } else {
-      self.dev_dependencies.swap_remove(&name);
-      self.dependencies.insert(name, version);
+    fn insert_index(object: &CstObject, searching_name: &str) -> usize {
+      object
+        .properties()
+        .into_iter()
+        .take_while(|prop| {
+          let prop_name =
+            prop.name().and_then(|name| name.decoded_value().ok());
+          match prop_name {
+            Some(current_name) => {
+              searching_name.cmp(&current_name) == std::cmp::Ordering::Greater
+            }
+            None => true,
+          }
+        })
+        .count()
     }
-  }
-
-  fn remove(&mut self, package: &str) -> bool {
-    let in_deps = self.dependencies.shift_remove(package).is_some();
-    let in_dev_deps = self.dev_dependencies.shift_remove(package).is_some();
-    in_deps || in_dev_deps
-  }
 
-  fn take_import_fields(
-    &mut self,
-  ) -> Vec<(&'static str, IndexMap<String, String>)> {
-    vec![
-      ("dependencies", std::mem::take(&mut self.dependencies)),
-      (
-        "devDependencies",
-        std::mem::take(&mut self.dev_dependencies),
-      ),
-    ]
-  }
-}
-
-struct NpmConfig {
-  config: Arc<deno_node::PackageJson>,
-  fmt_options: Option<FmtOptionsConfig>,
-  dependencies: IndexMap<String, String>,
-  dev_dependencies: IndexMap<String, String>,
-}
-
-enum DenoOrPackageJson {
-  Deno(DenoConfig),
-  Npm(NpmConfig),
-}
-
-impl From<DenoConfig> for DenoOrPackageJson {
-  fn from(config: DenoConfig) -> Self {
-    Self::Deno(config)
-  }
-}
-
-impl From<NpmConfig> for DenoOrPackageJson {
-  fn from(config: NpmConfig) -> Self {
-    Self::Npm(config)
-  }
-}
+    match self.kind {
+      ConfigKind::DenoJson => {
+        let imports = self.root_object.object_value_or_set("imports");
+        let value =
+          format!("{}@{}", selected.package_name, selected.version_req);
+        if let Some(prop) = imports.get(&selected.import_name) {
+          prop.set_value(json!(value));
+        } else {
+          let index = insert_index(&imports, &selected.import_name);
+          imports.insert(index, &selected.import_name, json!(value));
+        }
+      }
+      ConfigKind::PackageJson => {
+        let deps_prop = self.root_object.get("dependencies");
+        let dev_deps_prop = self.root_object.get("devDependencies");
+
+        let dependencies = if dev {
```

Comment:
Will this still remove the entry from the other "dependencies" when inserting it?

Ie. will `deno install npm:foo` and then `deno install --dev npm:foo` leave only a single entry and not entry in both `dependencies` and `devDependencies`?